src/perl.c

```diff
@@ -1008,11 +1011,24 @@ static int pplugin_call_all (pTHX_ int type, ...)
 {
 	int retvals = 0;
 
+	_Bool old_running;
 	va_list ap;
 	int ret = 0;
 
 	dSP;
 
+	c_ithread_t *t = (c_ithread_t *)pthread_getspecific(perl_thr_key);
+	if (t == NULL) /* thread destroyed ( c_ithread_destroy*() -> log_debug() ) */
+		return 0;
+
+	old_running = t->running;
+	t->running = 1;
+	
+	if (t->shutdown) {
+		t->running = old_running;
+		return 0;
+	}
+
 	if ((type < 0) || (type >= PLUGIN_TYPES))
```

Comment:
t->running needs to be reset before returning here. The same is true for fc_call() below. I think instead of your current approach, we should have a wrapper around call_pv(). That's where the actual switch from C to Perl happens and I don't think the Perl stack modifications that happen here before call_pv() would contribute to the problem you're fixing.